src/docs_rs.rs

```diff
@@ -0,0 +1,101 @@
+use async_trait::async_trait;
+use http::StatusCode;
+use mockall::automock;
+use url::Url;
+
+#[derive(Debug, thiserror::Error)]
+pub enum DocsRsError {
+    /// The rebuild couldn't be triggered.
+    /// The reason is passed in the given error message.
+    #[error("Bad request: {0}")]
+    BadRequest(String),
+    /// The request was rate limited by the server.
+    /// This is the NGINX level rate limit for requests coming from a single IP.
+    /// This is _not_ the rate limit that docs.rs might apply for rebuilds of the same crate
+    /// (AKA: "rebuild too often").
+    #[error("rate limited")]
+    RateLimited,
+    #[error("unauthorized")]
+    Unauthorized,
+    /// crate or version not found on docs.rs.
+    /// This can be temporary directly after a release until the first
+    /// docs build was started for the crate.
+    #[error("crate or version not found on docs.rs")]
+    NotFound,
+    #[error(transparent)]
+    Other(anyhow::Error),
+}
+
+#[automock]
+#[async_trait]
+pub trait DocsRsClient: Send + Sync {
+    async fn rebuild_docs(&self, name: &str, version: &str) -> Result<(), DocsRsError>;
+}
+
+pub(crate) struct RealDocsRsClient {
+    client: reqwest::Client,
+    base_url: Url,
+    api_token: String,
+}
+
+impl RealDocsRsClient {
+    pub fn new(base_url: impl Into<Url>, api_token: impl Into<String>) -> Self {
+        Self {
+            client: reqwest::Client::new(),
```

Comment:
```suggestion
            client: reqwest::Client::builder().user_agent("crates.io").build().unwrap(),
```

should we set a `User-Agent` header to make it easier for you to filter out non-crates.io traffic? :)

Reply:
good point, will do